homeassistant/components/climate/modbus.py

```diff
@@ -0,0 +1,173 @@
+    vol.Optional(CONF_DATA_TYPE, default=DATA_TYPE_FLOAT):
+        vol.In([DATA_TYPE_INT, DATA_TYPE_UINT, DATA_TYPE_FLOAT]),
+    vol.Optional(CONF_COUNT, default=2): cv.positive_int,
+    vol.Optional(CONF_PRECISION, default=1): cv.positive_int,
+    vol.Optional(CONF_MIN_TEMP): cv.positive_int,
+    vol.Optional(CONF_MAX_TEMP): cv.positive_int,
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = "mdi:thermometer"
+
+SUPPORT_FLAGS = SUPPORT_TARGET_TEMPERATURE
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Modbus Thermostat Platform."""
+    name = config.get(CONF_NAME)
+    modbus_slave = config.get(CONF_SLAVE)
+    target_temp_register = config.get(CONF_TARGET_TEMP)
+    current_temp_register = config.get(CONF_CURRENT_TEMP)
+    data_type = config.get(CONF_DATA_TYPE)
+    count = config.get(CONF_COUNT)
+    precision = config.get(CONF_PRECISION)
+    min_temp = config.get(CONF_MIN_TEMP)
+    max_temp = config.get(CONF_MAX_TEMP)
+
+    add_devices([ModbusThermostat(name, modbus_slave, target_temp_register,
+                                  current_temp_register, data_type, count,
+                                  precision, min_temp, max_temp)], True)
+
+
+class ModbusThermostat(ClimateDevice):
+    """Representation of a Modbus Thermostat."""
+
+    def __init__(self, name, modbus_slave, target_temp_register,
+                 current_temp_register, data_type, count, precision,
+                 min_temp, max_temp):
+        """Initialize the unit."""
+        self._name = name
+        self._slave = modbus_slave
+        self._target_temperature_register = int(target_temp_register)
+        self._current_temperature_register = int(current_temp_register)
+        self._target_temperature = None
+        self._current_temperature = None
+        self._data_type = data_type
+        self._count = int(count)
+        self._precision = precision
+        self._min_temp = min_temp
+        self._max_temp = max_temp
+        self._structure = '>f'
+
+        data_types = {DATA_TYPE_INT: {1: 'h', 2: 'i', 4: 'q'}}
+        data_types[DATA_TYPE_UINT] = {1: 'H', 2: 'I', 4: 'Q'}
+        data_types[DATA_TYPE_FLOAT] = {1: 'e', 2: 'f', 4: 'd'}
+
+        self._structure = '>{}'.format(data_types[self._data_type]
+                                       [self._count])
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return SUPPORT_FLAGS
+
+    def update(self):
+        """Update Target & Current Temperature."""
+        result = modbus.HUB.read_holding_registers(
+            self._slave, self._target_temperature_register, self._count)
+        byte_string = b''.join(
+            [x.to_bytes(2, byteorder='big') for x in result.registers])
+        val = struct.unpack(self._structure, byte_string)[0]
+        self._target_temperature = format(val, '.{}f'.format(self._precision))
+
+        result = modbus.HUB.read_holding_registers(
```

Comment:
Might be worth doing some debug logs on raw values received to help with troubleshooting later on. Unsure if a more specific error handling would be helpful.

Reply:
Yes I agree - I will add some more debugging info to this pull request.